File: src/lessons/welcome/traversal/SnakeEntity.c
//RemoteBuggle

int endingPosition();
void snakeStep() ;

/* BEGIN TEMPLATE */
void run() {
/* BEGIN SOLUTION */
	brushDown();
	while (!endingPosition()) {
		snakeStep();
	}
}
int endingPosition() {
	if (! isFacingWall())
		return 0;

	int res = 0;
	left();
	if (isFacingWall())
		res = 1;
	right();
	return res;
}

void snakeStep() {
	if (isFacingWall()) {
		if (getDirection() == EAST) {
			left();
			forward(1);
			left();
		} else {
			right();
			forward(1);
			right();
		}
	} else {
		forward(1);
	}

	/* END SOLUTION */
}
/* END TEMPLATE */

File: src/lessons/welcome/traversal/SnakeEntity.java
package lessons.welcome.traversal;

import plm.universe.Direction;
import plm.universe.bugglequest.SimpleBuggle;

public class SnakeEntity extends SimpleBuggle {
  @Override
  /* BEGIN TEMPLATE */
  public void run()
  {
    /* BEGIN SOLUTION */
    brushDown();
    while (!endingPosition()) {
      snakeStep();
    }
  }
  boolean endingPosition()
  {
    if (!isFacingWall())
      return false;

    boolean res = false;
    left();
    if (isFacingWall())
      res = true;
    right();
    return res;
  }

  void snakeStep()
  {
    if (isFacingWall()) {
      if (getDirection() == Direction.EAST) {
        left();
        stepForward();
        left();
      } else {
        right();
        stepForward();
        right();
      }
    } else {
      stepForward();
    }

    /* END SOLUTION */
  }
  /* END TEMPLATE */
}
